.teamcity/Gradle_Check/projects/FunctionalTestProject.kt

```diff
@@ -7,30 +7,27 @@ import model.CIBuildModel
 import model.Stage
 import model.TestCoverage
 
-class FunctionalTestProject(model: CIBuildModel, testConfig: TestCoverage, stage: Stage) : Project({
+class FunctionalTestProject(model: CIBuildModel, testConfig: TestCoverage, stage: Stage, deferredFunctionalTests: MutableList<FunctionalTest>) : Project({
     this.uuid = testConfig.asId(model)
     this.id = AbsoluteId(uuid)
     this.name = testConfig.asName()
-
-    model.buildTypeBuckets.forEach { bucket ->
+}) {
+    val functionalTests = model.buildTypeBuckets.flatMap { bucket ->
         if (bucket.shouldBeSkipped(testConfig) || !bucket.hasTestsOf(testConfig.testType)) {
-            return@forEach
+            return@flatMap emptyList<FunctionalTest>()
+        }
+        val functionalTestsForCurrentBucket = bucket.forTestType(testConfig.testType).map {
+            buildTypeBucket -> FunctionalTest(model, testConfig, buildTypeBucket.getSubprojectNames(), stage, buildTypeBucket.name, buildTypeBucket.extraParameters())
         }
         if (bucket.shouldBeSkippedInStage(stage)) {
-            addMissingTestCoverage(testConfig)
-            return@forEach
+            deferredFunctionalTests.addAll(functionalTestsForCurrentBucket)
```

Comment:
If I'm understanding correctly, this is modifying the passed parameter in `flatMap` (and caller's `map` [here](https://github.com/gradle/gradle/blob/859bd5b64aa7f3f613aac96b63d8094399221125/.teamcity/Gradle_Check/projects/StageProject.kt#L56)). Personally I don't like the side effect like this, can we somehow avoid it?

Reply:
Now the `flatMap` doesn't have a side-effect any more.